pkg/kubelet/images/pullmanager/image_pull_manager.go

```diff
@@ -0,0 +1,472 @@
+						klog.ErrorS(err, "failed to write an image pulled record", "image", image, "imageRef", imageRef)
+					}
+					return false
+				}
+			}
+		}
+	}
+
+	return true
+}
+
+func (f *PullManager) PruneUnknownRecords(imageList []string, until time.Time) {
+	f.pulledAccessors.GlobalLock()
+	defer f.pulledAccessors.GlobalUnlock()
+
+	pulledRecords, err := f.recordsAccessor.ListImagePulledRecords()
+	if err != nil {
+		klog.ErrorS(err, "failed to garbage collect ImagePulledRecords")
+		return
+	}
+
+	imagesInUse := sets.New(imageList...)
+	for _, ir := range pulledRecords {
+		if !ir.LastUpdatedTime.Time.Before(until) {
+			// the image record was only update after the GC started
+			continue
+		}
+
+		if imagesInUse.Has(ir.ImageRef) {
+			continue
+		}
+
+		if err := f.recordsAccessor.DeleteImagePulledRecord(ir.ImageRef); err != nil {
+			klog.ErrorS(err, "failed to remove an ImagePulledRecord", "imageRef", ir.ImageRef)
+		}
+	}
+
+}
+
+// initialize gathers all the images from pull intent records that exist
+// from the previous kubelet runs.
+// If the CRI reports any of the above images as already pulled, we turn the
+// pull intent into a pulled record and the original pull intent is deleted.
+//
+// This method is not thread-safe and it should only be called upon the creation
+// of the PullManager.
+func (f *PullManager) initialize(ctx context.Context) {
+	pullIntents, err := f.recordsAccessor.ListImagePullIntents()
+	if err != nil {
+		klog.ErrorS(err, "there was an error listing ImagePullIntents")
+		return
+	}
+
+	if len(pullIntents) == 0 {
+		return
+	}
+
+	images, err := f.imageService.ListImages(ctx)
+	if err != nil {
+		klog.ErrorS(err, "failed to list images")
+	}
+
+	inFlightPulls := sets.New[string]()
+	for _, intent := range pullIntents {
+		inFlightPulls.Insert(intent.Image)
+	}
+
+	// Each of the images known to the CRI might consist of multiple tags and digests,
+	// which is what we track in the ImagePullIntent - we need to go through all of these
+	// for each image.
+	for _, image := range images {
+		imageName := searchForExistingTagDigest(inFlightPulls, image)
+		if len(imageName) == 0 {
+			continue
+		}
+
+		if err := f.writePulledRecord(imageName, image.ID, nil); err != nil {
+			klog.ErrorS(err, "failed to write an image pull record", "imageRef", image.ID)
+			continue
+		}
+
+		if err := f.recordsAccessor.DeleteImagePullIntent(imageName); err != nil {
+			klog.V(2).InfoS("failed to remove image pull intent file", "imageName", imageName, "error", err)
+		}
+
+	}
+}
+
+// searchForExistingTagDigest loop through the `image` RepoDigests and RepoTags
+// and tries to find a digest/tag in `trackedImages`, which is a map of
+// containerImage -> pulling intent path.
+func searchForExistingTagDigest(trackedImages sets.Set[string], image kubecontainer.Image) string {
```

Comment:
make it clear what the string this returns is (`image` spec or `imageRef`)

It's possible for there to be *multiple* image pull intents for differently named references to the same underlying imageRef, so this needs to return `images []string`

(also, rename trackedImages to inFlightPulls)